### ISSUE TITLE
Observer is stored and swiftlint still triggering a violation

### ISSUE THREAD

**Issue by BilalReffas** [realm/SwiftLint#2873]
### New Issue Checklist

- [ ] Updated SwiftLint to the latest version
- [ ] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

We store the Observer and swiftlint is still triggering a `discarded_notification_center_observer` violation.
##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
* Paste your configuration file:

```yml
enabled_rules:
  - private_action
  - private_outlet
disabled_rules:
  - trailing_comma
  - empty_enum_arguments
  - void_return
  - redundant_optional_initialization
  - nesting
  - redundant_discardable_let
```

```swift
// This triggers a violation:
        self.observations = ApplicationStateWatchdog.handledNotifications.map { notification in
            self.notificationCenter.addObserver(
                forName: notification,
                object: nil,
                queue: nil,
                using: self.handleStateUpdate
            )
        }
```

**Comment by marcelofabri** [comment]
Duplicate of #1398
